File: sdk/node/src/dsn-parser.ts
/**
 * DSN Parser for WorkerSQL
 * Parses connection strings in the format:
 * workersql://[username[:password]@]host[:port][/database][?param1=value1&param2=value2]
 */

export interface ParsedDSN {
  protocol: string;
  username?: string;
  password?: string;
  host: string;
  port?: number;
  database?: string;
  params: Record<string, string>;
}

export class DSNParser {
  /**
   * Parse a WorkerSQL DSN string
   * @param dsn Connection string to parse
   * @returns Parsed DSN components
   */
  static parse(dsn: string): ParsedDSN {
    if (!dsn || typeof dsn !== 'string') {
      throw new Error('DSN must be a non-empty string');
    }

    // Match the DSN pattern
    const dsnRegex = /^([a-z]+):\/\/(?:([^:@]+)(?::([^@]+))?@)?([^/:?]+)(?::(\d+))?(?:\/([^?]+))?(?:\?(.+))?$/i;
    const match = dsn.match(dsnRegex);

    if (!match) {
      throw new Error(`Invalid DSN format: ${dsn}`);
    }

    const [, protocol, username, password, host, portStr, database, queryString] = match;

    // Validate protocol
    if (!protocol || protocol.toLowerCase() !== 'workersql') {
      throw new Error(`Invalid protocol: ${protocol || 'missing'}. Expected 'workersql'`);
    }

    if (!host) {
      throw new Error('Host is required in DSN');
    }

    // Parse port
    const port = portStr ? parseInt(portStr, 10) : undefined;
    if (port !== undefined && (isNaN(port) || port < 1 || port > 65535)) {
      throw new Error(`Invalid port: ${portStr}`);
    }

    // Parse query parameters
    const params: Record<string, string> = {};
    if (queryString) {
      const pairs = queryString.split('&');
      for (const pair of pairs) {
        const [key, value] = pair.split('=');
        if (key) {
          params[decodeURIComponent(key)] = value ? decodeURIComponent(value) : '';
        }
      }
    }

    return {
      protocol,
      username: username ? decodeURIComponent(username) : undefined,
      password: password ? decodeURIComponent(password) : undefined,
      host: decodeURIComponent(host),
      port,
      database: database ? decodeURIComponent(database) : undefined,
      params,
    };
  }

  /**
   * Convert parsed DSN back to string format
   * @param parsed Parsed DSN components
   * @returns DSN string
   */
  static stringify(parsed: ParsedDSN): string {
    let dsn = `${parsed.protocol}://`;

    if (parsed.username) {
      dsn += encodeURIComponent(parsed.username);
      if (parsed.password) {
        dsn += `:${encodeURIComponent(parsed.password)}`;
      }
      dsn += '@';
    }

    dsn += encodeURIComponent(parsed.host);

    if (parsed.port) {
      dsn += `:${parsed.port}`;
    }

    if (parsed.database) {
      dsn += `/${encodeURIComponent(parsed.database)}`;
    }

    if (Object.keys(parsed.params).length > 0) {
      const queryString = Object.entries(parsed.params)
        .map(([key, value]) => `${encodeURIComponent(key)}=${encodeURIComponent(value)}`)
        .join('&');
      dsn += `?${queryString}`;
    }

    return dsn;
  }

  /**
   * Extract API endpoint from DSN parameters or construct from host
   * @param parsed Parsed DSN components
   * @returns API endpoint URL
   */
  static getApiEndpoint(parsed: ParsedDSN): string {
    // Check if apiEndpoint is specified in params
    if (parsed.params['apiEndpoint']) {
      return parsed.params['apiEndpoint'];
    }

    // Construct from host
    const protocol = parsed.params['ssl'] === 'false' ? 'http' : 'https';
    const port = parsed.port ? `:${parsed.port}` : '';
    return `${protocol}://${parsed.host}${port}/v1`;
  }
}

File: sdk/java/src/main/java/com/workersql/sdk/util/DSNParser.java
package com.workersql.sdk.util;

import java.io.UnsupportedEncodingException;
import java.net.URLDecoder;
import java.net.URLEncoder;
import java.util.HashMap;
import java.util.Map;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

/**
 * DSN Parser for WorkerSQL
 * Parses connection strings in the format:
 * workersql://[username[:password]@]host[:port][/database][?param1=value1&param2=value2]
 */
public class DSNParser {
    private static final Pattern DSN_PATTERN = Pattern.compile(
        "^([a-z]+)://(?:([^:@]+)(?::([^@]+))?@)?([^/:?]+)(?::(\\d+))?(?:/([^?]+))?(?:\\?(.+))?$",
        Pattern.CASE_INSENSITIVE
    );

    /**
     * Parse a WorkerSQL DSN string
     * @param dsn Connection string to parse
     * @return Parsed DSN components
     */
    public static ParsedDSN parse(String dsn) {
        if (dsn == null || dsn.isEmpty()) {
            throw new IllegalArgumentException("DSN must be a non-empty string");
        }

        Matcher matcher = DSN_PATTERN.matcher(dsn);
        if (!matcher.matches()) {
            throw new IllegalArgumentException("Invalid DSN format: " + dsn);
        }

        String protocol = matcher.group(1);
        String username = matcher.group(2);
        String password = matcher.group(3);
        String host = matcher.group(4);
        String portStr = matcher.group(5);
        String database = matcher.group(6);
        String queryString = matcher.group(7);

        // Validate protocol
        if (protocol == null || !protocol.equalsIgnoreCase("workersql")) {
            throw new IllegalArgumentException("Invalid protocol: " + protocol + ". Expected 'workersql'");
        }

        if (host == null || host.isEmpty()) {
            throw new IllegalArgumentException("Host is required in DSN");
        }

        // Parse port
        Integer port = null;
        if (portStr != null) {
            try {
                port = Integer.parseInt(portStr);
                if (port < 1 || port > 65535) {
                    throw new IllegalArgumentException("Invalid port: " + portStr);
                }
            } catch (NumberFormatException e) {
                throw new IllegalArgumentException("Invalid port: " + portStr);
            }
        }

        // Parse query parameters
        Map<String, String> params = new HashMap<>();
        if (queryString != null && !queryString.isEmpty()) {
            String[] pairs = queryString.split("&");
            for (String pair : pairs) {
                int idx = pair.indexOf('=');
                if (idx > 0) {
                    String key = urlDecode(pair.substring(0, idx));
                    String value = idx < pair.length() - 1 ? urlDecode(pair.substring(idx + 1)) : "";
                    params.put(key, value);
                }
            }
        }

        return new ParsedDSN(
            protocol,
            username != null ? urlDecode(username) : null,
            password != null ? urlDecode(password) : null,
            urlDecode(host),
            port,
            database != null ? urlDecode(database) : null,
            params
        );
    }

    /**
     * Convert parsed DSN back to string format
     * @param parsed Parsed DSN components
     * @return DSN string
     */
    public static String stringify(ParsedDSN parsed) {
        StringBuilder dsn = new StringBuilder(parsed.getProtocol()).append("://");

        if (parsed.getUsername() != null) {
            dsn.append(urlEncode(parsed.getUsername()));
            if (parsed.getPassword() != null) {
                dsn.append(":").append(urlEncode(parsed.getPassword()));
            }
            dsn.append("@");
        }

        dsn.append(urlEncode(parsed.getHost()));

        if (parsed.getPort() != null) {
            dsn.append(":").append(parsed.getPort());
        }

        if (parsed.getDatabase() != null) {
            dsn.append("/").append(urlEncode(parsed.getDatabase()));
        }

        if (!parsed.getParams().isEmpty()) {
            dsn.append("?");
            boolean first = true;
            for (Map.Entry<String, String> entry : parsed.getParams().entrySet()) {
                if (!first) dsn.append("&");
                dsn.append(urlEncode(entry.getKey())).append("=").append(urlEncode(entry.getValue()));
                first = false;
            }
        }

        return dsn.toString();
    }

    /**
     * Extract API endpoint from DSN parameters or construct from host
     * @param parsed Parsed DSN components
     * @return API endpoint URL
     */
    public static String getApiEndpoint(ParsedDSN parsed) {
        // Check if apiEndpoint is specified in params
        if (parsed.getParams().containsKey("apiEndpoint")) {
            return parsed.getParams().get("apiEndpoint");
        }

        // Construct from host
        String protocol = "false".equalsIgnoreCase(parsed.getParams().get("ssl")) ? "http" : "https";
        String port = parsed.getPort() != null ? ":" + parsed.getPort() : "";
        return protocol + "://" + parsed.getHost() + port + "/v1";
    }

    private static String urlDecode(String str) {
        try {
            return URLDecoder.decode(str, "UTF-8");
        } catch (UnsupportedEncodingException e) {
            throw new RuntimeException("UTF-8 encoding not supported", e);
        }
    }

    private static String urlEncode(String str) {
        try {
            return URLEncoder.encode(str, "UTF-8");
        } catch (UnsupportedEncodingException e) {
            throw new RuntimeException("UTF-8 encoding not supported", e);
        }
    }
}
